### PR TITLE
Update to new crates; simplify advanced options

### PR DESCRIPTION
**Pull request by ivmarkov**
Following changes:
* All dependencies raised to the newly-released `esp-idf-*` crates
* The advanced options simplified (we had too many). We now have basically "HAL?" y/n and then if "HAL" is selected, "STD?" y/n. Selecting neither HAL nor STD generates:
  * an `app_main` no-op Rust crate with the `cargo` driver 
  * a no-dependencies Rust crate with the `cmake` driver which might be useful for Rust code which is purely computational
* Simplified dependency tree - only `esp-idf-svc`, `log` and `embuild` (the last as a build dependency); no dependencies when the HAL option is not enabled

### CODE REVIEW

**Comment by ivmarkov** [comment]
Still in draft as I need to address all build failures.

**Comment by ivmarkov** [comment]
Also pulled-in the article of @georgik from [here](https://github.com/georgik/esp32-idf-no-std-rust-component). 

There's a slight difference between the project layout described in the article (Rust lib crate source code is separate from the ESP IDF component C source code) versus what is generated by the template (Rust and C source code mixed in a single, hybrid, ESP-IDF-component-but-also-a-rust-lib-crate thing) which might have to be addressed too.

**Comment by Vollbrecht** [comment]
Not directly related to the current changes but i want to bring it up while we make changes here. I think a nice addition would be to create two sdkconfig.defaults. I think the option to have two different sdkconfig for release and debug is a good usecase, but the option is in my opinion not visible enough. A good difference between the two would be to set different CONFIG_LOG_MAXIMUM_LEVEL in each. In my simple wifi only test app i shaved 100k flash usage just in strings from this settings. The idea of adding this is merely to promote more, that this options exist.

**Comment by ivmarkov** [comment]
> A good difference between the two would be to set different CONFIG_LOG_MAXIMUM_LEVEL in each. In my simple wifi only test app i shaved 100k flash usage just in strings from this settings. The idea of adding this is merely to promote more, that this options exist.

Can you show an example how you achieved that?
In general, I think the template needs simplification rather than more bells and whistles, so doing two sdkconfig files has to have a really compelling use case.

In fact, we might be better of with a sinbgle file, and then a bunch of useful options either enabled there, or disabled via a comment.

**Comment by Vollbrecht** [comment]
I think creating a well configured project and files in it, is a different goal than make the generation process less complex for the user. I am on board if you say that there might be two many options for the user, but that doesn't mean that the setup should not generate "labor intensive" things that the user might want to do. Also we explicitly have a fast path for most users - using the "non-advanced" branch.  I mean we labeled them "advanced" options for a reason. Just cutting things and let the user handle the task themself is not an improvement.

In the case of of sdkconfigs. We can have two files : `sdkconfig.debug` and `sdkconfig.release`. This naming scheme should work across all different mcu's so we dont need multiple per mcu target. What we should set there as a sane default is a question up for debate, but we should outline at least all the common configs that we are highly likely to encounter often.

Setting different max log levels is just one low hanging fruit. Even if we don't change it up and only have one config file, we should include more settings ( out-commented ) like the different stack_size options for event_task etc / https /wss related / ( uart - usb serial related) and similar. 

This changes would not create a more or less complex cargo generate matrix. I can try to make a follow up PR with the related changes when this is finished here.

**Comment by ivmarkov** [comment]
I'm all for specifically putting some extra options in `sdkconfig.defaults`. Or - if really necessary - having two separate such files.

I do not agree that the purpose of the template is to automate labor intensive tasks and become a swiss army knife.
Its purpose is to get (primarily newbie) users up to speed, with something which does build out of the box, and that's it.

Which is also the original purpose of `cargo new`, I believe - the tool which we are approaching here.

**Comment by ivmarkov** [comment]
@SergioGasquez Would appreciate your review. Summary of changes:
* `ci_cargo` - I've unified the v4 and v5 build to a single job, so as to avoid copy-pasting across the two
* `ci_cmake` - Support for all MCUs
* The CMake docu now has a new "how it works" article. Initially, I used the one from georgik, but then reworked it so as not be a step-by-step guide, but rather to explain how the CMake+Cargo build works under the hood
* Most importantly, the template now works with the latest versions of all `esp-idf-*` crates

File: README-cmake-details.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+This is a pretty standard and mostly empty `cargo` [build manifest file](https://doc.rust-lang.org/cargo/reference/manifest.html), which sets up your Rust crate to be built as a Rust static libray. 
+
+If you have selected the "HAL" option when generating the project, the manifest would also list:
+* a dependency to `esp-idf-svc` (i.e. [the safe Rust wrappers for ESP-IDF](https://github.com/esp-rs/esp-idf-svc));
+* a built-time dependency to [`embuild`](https://github.com/esp-rs/embuild) which helps the build integration between ESP-IDF and Rust to run smoothly);
+* the popular Rust [`log`](https://github.com/rust-lang/log) crate.
+
+Post-generation, you can add/remove additional dependencies on Rust crates here as well as modify how Rust compiles your project (as in optimization level and others).
+
+#### `components/rust-test/rust-toolchain.toml`:
```
Comment (SergioGasquez) [inline README-cmake-details.md:104]: Maybe link to https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file?

File: README-cmake-details.md [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+## How the Build Process Works
+
+The CMake build for your `rust-test` component would proceed as usual and will compile all C code in the component (which is not much - just an empty `placeholder.c` file which needs to be there for the approach to work).
+
+More importantly, `components/rust-test/CMakeLists.txt` is setup by the generator in such a way, so as to **also** call the Rust standard build utility - [`cargo`](https://doc.rust-lang.org/cargo/index.html) - which would take care of compiling the Rust source code in your ESP-IDF component.
+
+The Rust source code (or as it is named in the Rust world - the Rust *crate*) is setup - in `components/rust-test/Cargo.toml` - to have the shape of a *static* library - just like any other ESP-IDF component. When the CMake build calls `cargo` to compile your `rust-test` Rust crate, you'll essentially end up with a static library named `librust_test.a`, which is then linked to the rest of the ESP-IDF project by CMake.
+
+## Project Structure Details
```
Comment (SergioGasquez) [inline README-cmake-details.md:53]: Maybe place this section under "Generated Project Structure" or at least, after it.

File: README-cmake-details.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+* the popular Rust [`log`](https://github.com/rust-lang/log) crate.
+
+Post-generation, you can add/remove additional dependencies on Rust crates here as well as modify how Rust compiles your project (as in optimization level and others).
+
+#### `components/rust-test/rust-toolchain.toml`:
+
+This file instructs `cargo` what Rust compiler toolchain to use when building your component. It would list either the `esp` toolchain (a Rust toolchain provided by Espressif that has extra support for the esp32 xtensa MCUs), or the standard Rust `nightly` toolchain. The latter can only be used with Espressif RiscV MCUs.
+
+#### `components/rust-test/src/lib.rs`:
+The actual source code of your Rust library. You can extend here at will.
```
Comment (SergioGasquez) [inline README-cmake-details.md:109]:
```suggestion
#### `components/rust-test/src/lib.rs`:

The actual source code of your Rust library. You can extend here at will.
```

File: README-cmake-details.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+Create [`wokwi.toml`](./wokwi.toml) in the root of the project. The file contains references to BIN and ELF previously built by `idf.py`.
+
+```toml
+[wokwi]
+version = 1
+elf = "build/test.elf"
+firmware = "build/test.bin"
+```
+
+Create [`diagram.json`](./diagram.json). The file contains board selected for the simulation.
```
Comment (SergioGasquez) [inline README-cmake-details.md:136]: Maybe link to the available boards? https://github.com/wokwi/wokwi-boards/tree/main/boards

File: cargo/src/main.rs [hunk 5]
```diff
@@ -1,25 +1,18 @@
...
-{%- if std and hal != "No" %}
-use log::*;
-{% endif %}
 
-{%- unless std %}
+{% endunless -%}
+{% unless std -%}
 #[no_mangle]
-{%- endunless %}
+{% endunless -%}
```
Comment (SergioGasquez) [inline cargo/src/main.rs:8]: Cant this conditions be joined?

**Review by SergioGasquez — APPROVED** [review]
Thanks for this huge work! Just left a few details in the comments. 

Also, it would be good if we find a way to skip the run of impossible CI cases (eg: C6 with esp-idf v4.4) without complicating too much the CI files (ie: duplicating code or adding lots of ifs)

File: README-cmake-details.md (hunk 2 above)
Comment (ivmarkov) [reply README-cmake-details.md:53]: Moved the "How The build Works" as a subsection of "Project Structure".

**Comment by ivmarkov** [comment]
> Thanks for this huge work! Just left a few details in the comments.
> 
> Also, it would be good if we find a way to skip the run of impossible CI cases (eg: C6 with esp-idf v4.4) without complicating too much the CI files (ie: duplicating code or adding lots of ifs)

I agree, good idea, but maybe next time? I barely managed to get this one working.

File: cargo/src/main.rs (hunk 5 above)
Comment (ivmarkov) [reply cargo/src/main.rs:8]: Same here. Don't want to touch it too much. biggest hurdle is to get the whitespacing right so `cargo fmt` does not complain.

**Pull request merged**